internal/types/subtypes/interceptor.go

```diff
@@ -0,0 +1,193 @@
+			return req
+		}
+	case idField != nil && attributesField != nil:
+		id := r.Get(idField).String()
+		st = SubtypeFromId(domain, id)
+		if err := convertAttributesToSubtype(req, st); err != nil {
+			return req
+		}
+	}
+	return req
+}
+
+func transformResponseItemAttributes(item proto.Message) error {
+	r := item.ProtoReflect()
+
+	itemFields := r.Descriptor().Fields()
+	typeField := itemFields.ByName("type")
+	if typeField == nil {
+		// not an item with subtypes
+		return nil
+	}
+
+	st := Subtype(item.ProtoReflect().Get(typeField).String())
+	return convertAttributesToDefault(item, st)
+}
+
+func transformResponseAttributes(_ context.Context, res proto.Message) proto.Message {
+	r := res.ProtoReflect()
+	fields := r.Descriptor().Fields()
+
+	itemField := fields.ByName("item")
+	itemsField := fields.ByName("items")
+	switch {
+	case itemField != nil:
+		if itemR := itemField.Message(); itemR == nil {
+			return res
+		}
+
+		item := r.Get(itemField).Message().Interface()
+		if err := transformResponseItemAttributes(item); err != nil {
+			return res
+		}
+	case itemsField != nil:
+		if !itemsField.IsList() {
+			return res
+		}
+		items := r.Get(itemsField).List()
+
+		for i := 0; i < items.Len(); i++ {
+			item := items.Get(i).Message().Interface()
+			if err := transformResponseItemAttributes(item); err != nil {
+				return res
+			}
+		}
+
+	}
+	return res
+}
+
+// AttributeTransformerInterceptor is a grpc server interceptor that will
+// transform subtype attributes for requests and responses. This will only
+// modify requests and responses that adhere to a specific structure and is done
+// to support the use of a oneof for attributes to strongly type the attributes
+// while allowing the JSON API to provide attributes via a single key.
+//
+// For example with a protobuf message definition like:
+//
+//     message Account {
+//       string id = 1;
+//       string type = 2;
+//       oneof attrs {
+//         google.protobuf.Struct attributes = 10 [(controller.custom_options.v1.subtype) = "default"];
+//         PasswordAttributes password_attributes = 20 [(controller.custom_options.v1.subtype) = "password"];
+//       }
+//     }
+//
+//     message PasswordAttributes {
+//        string login_name = 1;
+//     }
+//
+//     message AccountCreateRequest {
+//        Account item = 1;
+//     }
+//
+//     message AccountCreateResponse {
+//        Account item = 1;
+//     }
+//
+// And a create request with JSON request body like:
+//     {
+//        "type": "password",
+//        "attributes": {
+//           "login_name": "tim"
+//        }
+//     }
+//
+// Will result in a proto request like:
+//
+//     type:"password" attributes:{fields:{key:"login_name" value:{string_value:"tim"}}}
+//
+// This request will be transformed into:
+//
+//     type:"password" password_attributes:{login_name:"tim"}
+func AttributeTransformerInterceptor(_ context.Context) grpc.UnaryServerInterceptor {
```

Comment:
Do we need the context parameter?

Reply:
The other [interceptors](https://github.com/hashicorp/boundary/blob/tmessi-subtype-grpc-interceptor/internal/servers/controller/gateway.go) all take a context. I think it is likely that this context will be needed or used at some point, since we need a context to make error events. So I think it should be included even if it is not used yet.

Reply:
They are not scoped to the request context. If you look at the other interceptors, most take a context but do not use it, the ones that do use it are performing [initialization checks](https://github.com/hashicorp/boundary/blob/dd988c1d867c813759a5aa0f2fb1151c69791cf6/internal/servers/controller/interceptor.go#L55-L73). I don't know what initialization checks could be added to this interceptor, but I think consistency with the other interceptors is a good idea here.

Reply:
That sounds reasonable 👍🏻, thanks for explaining.